Save logging + Fixes

PsyCommando opened the pull request:
<!-- !! PLEASE, READ THIS !! -->
<!-- We recommend to check the contributing page before opening pull requests. -->
<!-- https://github.com/NebulaSS13/Nebula/blob/dev/CONTRIBUTING.md -->
<!-- If you're opening a pull request which changes A LOT of icon/map files: -->
<!-- Add [IDB IGNORE] (to ignore icon file changes) or [MDB IGNORE] (to ignore map file changes) in the PR title. -->
<!-- These tags prevent huge diffs from overloading IconDiffBot and MapDiffBot. -->

## Description of changes
Fixes:
* Fix z level issues with save loading
* Fix planetoid data not restoring correctly from save.

Changes:
* Logs each saves to db.
* Add some db side procedures.
* Cut up loading and saving code into smaller bits so it's a bit more manageable.
* Added some customizable error tolerance to saving/loading.
* Critical errors handling.
* Recoverable errors handling.
* Added an admin verb under server to pick what kind of errors to tolerate during saving/loading.

PsyCommando committed:
Add save logging + cutup save code

PsyCommando committed:
Save error tolerance admin verb

Added admin verb to the server tab to allow changing at runtime the error tolerance for saving the world. Allow skipping over recoverable and unrecoverable error if wanted. Meant to be a sort of last-resort tool to force a save if something on the map breaks the save in a limited way.

PsyCommando committed:
Update misc.dm

PsyCommando committed:
Reorganized and moved saving code

Renamed and moved the code for saving in SSpersistence over to it's own file since it was pretty massive and hard to navigate.

PsyCommando committed:
Update persistence.dm

PsyCommando committed:
Moved save statistics to own file

PsyCommando committed:
Moved limbo stuff to it's own file

* Also renamed "DeserializeOneOff" to "LoadFromLimbo" to match the other limbo proc names.

PsyCommando committed:
Began moving save loading code to new file

PsyCommando committed:
Update persistence.dm

PsyCommando committed:
Fix mimic_edge saving

PsyCommando committed:
fixed save text being hard to read in darkmode

PsyCommando committed:
Moved and cut up world loading code

* Renamed and duplicated exception filters, one for saving, and one for loading. Since loading doesn't need to unblock entering or restart subsystems on an exception.
* Cut up loading code after applying some of the fixes upstream.

PsyCommando committed:
TEMPORARY fix for debugger not working

PsyCommando committed:
Formatting

PsyCommando committed:
Fix map not loading at the right zlevel

PsyCommando committed:
Make planetoid_data load properly

PsyCommando committed:
Update check-paths.sh

PsyCommando committed:
Revert "TEMPORARY fix for debugger not working"

This reverts commit cee6ea4a90d44022c6faf20ea686eee9cad8dac2.

PsyCommando committed:
Move last save display to persistent modpack

NataKilar commented on mods/persistence/__defines/serializer.dm:
```diff
@@ -1,35 +1,65 @@
+// SQL Helpers
+/////////////////////////////////////////////////////////
+
+///A helper for executing a sql query and throwing the proper exception with a standardized error message. QUERY must be a variable.
```
The comment ``QUERY must be a variable`` here means ``QUERY must be an instance of DBQuery``?

PsyCommando replied:
Yep. But more accurately it means is it cannot be a statement. Otherwise it'll error out because there's no return types in byond.

NataKilar commented on mods/persistence/controllers/subsystems/autosave.dm:
```diff
@@ -22,53 +22,73 @@ SUBSYSTEM_DEF(autosave)
 		if(check_for_restart && reset_after_save)
-			to_world("<font size=4 color='red'>Server is resetting after this save!</font>")
+			to_world(SPAN_AUTOSAVE_WARN("Server is restarting after this autosave!"))
+		sleep(5)
```
What's the purpose of this sleep?

PsyCommando replied:
The server needs a bit of time to push messages to all connected users. Otherwise you get instances where nothing is shown on some of the clients because byond doesn't flush buffered messages.

NataKilar commented on mods/persistence/controllers/subsystems/persistence.dm:
```diff
@@ -1,679 +1,278 @@
+
+	//#FIXME: The idea for having a generic serializer class was that we wouldn't initiate the exact subtype in the var definition.
+	//        So that you could just interchangeably use another kind of serializer that saves to json or xml or whatever instead of sql seamlessly.
+	//        This implementation completely nullifies the whole point of having a serializer class currently?
```
Theoretically this is true, but I doubt we'll ever reach the point where the implementation is completely generic. Due to proc call overhead, there's also the chance that over generifying things will hit performance. There is definitely repeated code in the subtypes though that should be fixed.

PsyCommando replied:
Yeah, we're not going to have a need for something like that. But it's just food for thoughts.

NataKilar commented on mods/persistence/controllers/subsystems/persistence.dm:
```diff
@@ -1,679 +1,278 @@
-
-/datum/controller/subsystem/persistence/proc/print_db_status()
-	return SQLS_Print_DB_STATUS()
+/mob/Stat()
```
What is this doing, and why is it here?

PsyCommando replied:
It's printing the last save loaded to admins/server staff in their status panel. It's pretty well documented in the doc.

NataKilar replied:
Sorry, I'm referring to the ``/mob/Stat()`` proc below.

PsyCommando replied:
That's what I was replying to.

NataKilar commented on mods/persistence/controllers/subsystems/persistence/persistence_loading.dm:
```diff
@@ -0,0 +1,297 @@
+
+				var/datum/V
+				try
+					V = _list[key] //#FIXME: We really need to get rid of this awful way to check list types.
```
Agreed.

NataKilar commented on mods/persistence/controllers/subsystems/persistence/persistence_loading.dm:
```diff
@@ -0,0 +1,297 @@
+			_handle_recoverable_load_exception(e, "while updating off-world storage state for player '[char_mind.key]'")
+
+///Deserialize cached turfs exclusively from the db cache.
+/datum/controller/subsystem/persistence/proc/_deserialize_turfs()
```
This proc name is slightly inaccurate, since it can deserialize wrapper holders as well. It might be worth naming it something like "_deserialize_root_datums" since technically *everything* will be deserialized through the contents chain. Not a big deal either way.

PsyCommando replied:
I'll leave it like that for now.

NataKilar commented on mods/persistence/controllers/subsystems/persistence/persistence_loading.dm:
```diff
@@ -0,0 +1,297 @@
+		for(var/turf/T in block(locate(1, 1, z_level.new_index), locate(world.maxx, world.maxy, z_level.new_index)))
+			try
+				if(current_area)
+					current_area.contents += T //#FIXME: It's dangerous to do it like this. Use loc on the turf, not the area's contents "list".
```
You can make this change now if you'd like.

PsyCommando replied:
I'm just leaving comments for later. This isn't the PR to make changes like that.

NataKilar commented on mods/persistence/controllers/subsystems/persistence/persistence_loading.dm:
```diff
@@ -0,0 +1,297 @@
+				// Adjust to new index.
+				coords[3] = serializer.z_map[coords[3]]
+				var/turf/T = locate(text2num(coords[1]), text2num(coords[2]), coords[3])
+				new_area.contents += T //#FIXME: Accessing contents directly is dangerous. It's better to set loc instead.
```
Ditto.

PsyCommando replied:
Same as last reply.

NataKilar commented on mods/persistence/controllers/subsystems/persistence/persistence_loading.dm:
```diff
@@ -0,0 +1,297 @@
+	report_progress_serializer("Deserialized [LAZYLEN(turfs_loaded)] turfs and their contents in [REALTIMEOFDAY2SEC(time_start)]s.")
+	sleep(5)
+
+/// TODO
```
Todo?

PsyCommando replied:
The doc for the proc is todo.

NataKilar commented on mods/persistence/controllers/subsystems/persistence/persistence_saving.dm:
```diff
@@ -0,0 +1,437 @@
+		S.disable()
+
+	//Wait on SSair to complete it's tick.
+	if (SSair.state != SS_IDLE)
```
Will disabling the subsystems above stop SSair from completing its tick? Also ``it's -> its`` if you get a chance

PsyCommando replied:
Disable does exactly the same thing it did before.
```
/datum/controller/subsystem/proc/disable()
	can_fire = FALSE
```

NataKilar commented on mods/persistence/controllers/subsystems/persistence/persistence_saving.dm:
```diff
@@ -0,0 +1,437 @@
+	//!! - HOT CODE BELOW - !!
+	//!!!!!!!!!!!!!!!!!!!!!!!!
+	for(var/z in saved_levels)
+		var/datum/persistence/load_cache/z_level/z_level = z_transform["[z]"] //#FIXME: String concatenation are extremely slow!!!
```
True, but this is only in the ``saved_levels`` loop. Creating the string as a key here isn't going to cause an issue.

PsyCommando replied:
fixed

NataKilar commented on mods/persistence/controllers/subsystems/persistence/persistence_saving.dm:
```diff
@@ -0,0 +1,437 @@
+						if(last_area_type != TA.type || last_area_name != TA.name)
+							if(area_turf_count > 0)
+								z_level.areas += list(list("[last_area_type]", sanitize_sql(last_area_name), area_turf_count))
+							last_area_type = TA.type //#FIXME: If last_area_type is only ever used as a string, might as well concat it here, instead of doing it thousands of time above?
```
The string concatenation should only occur at the boundaries of two areas, when ``last_area_type`` changes. Changing ``last_area_type`` to a string would be slower, as you'd need to convert ``TA.type`` to a string for comparison, which *would* occur thousands of times.

PsyCommando replied:
alright

NataKilar commented on mods/persistence/modules/overmap/sectors.dm:
```diff
@@ -3,7 +3,7 @@
 	var/atom/old_loc	 	 // Where the ship was prior to saving. Used to relocate the ship following saving, not on load.
 
-	//#FIXME: Don't store this here.
+	//#FIXME: Don't store this here. Those are meant to be strictly just temporary objects just to show a position on the overmap!
```
I don't think these should be regarded as temporary, since ship velocity etc. are stored on this object.

PsyCommando replied:
The thing is, that's what neb has been going for last time I asked. That's part of why most of the planet code was decoupled from the markers.

NataKilar replied:
Not blocking, but it's not like Nebula has saving either. There's nothing in game that creates and destroys these objects, so I'm not sure how they could be regarded as temporary.

NataKilar commented on mods/persistence/modules/world_save/save_testing.dm:
```diff
@@ -0,0 +1,27 @@
+	. = ..()
+	CRASH("obj/debug/error_on_save: Crashing save!")
+
+//
```
Missing comment?

PsyCommando committed:
Fix loading save not filling saved_levels

level_data.setup_level_data() is not called after loading level data from a save. But that's where it was adding itself to the list of saved levels. (which really should have stayed inside ssmapping for this very reason)

PsyCommando committed:
Removed duplicate cleanup

PsyCommando committed:
Throw the save/load exception if any

Silently catching it means we don't get a call stack and nice things like that. So, I'm rethrowing it for the runtime log to pick it up.

PsyCommando committed:
Cleanup save debug objects a bit

PsyCommando committed:
changed comments

PsyCommando committed:
Added configuration hooks for persistence configs

* Added the persistence config options to the example config file.
* Renamed an ambiguous config options for persistence.

PsyCommando commented:
Added some more things while waiting. Now persistence config options are actually parsed from the config files via hooks to the modpack.

PsyCommando committed:
Used a stub system for config

Removes dependency on the persistence modpack.

PsyCommando committed:
updated config example

PsyCommando committed:
Clarified comment on autosave config
